crates/wit-component/src/linking.rs

```diff
@@ -0,0 +1,1383 @@
+        for name in &metadata.table_address_imports {
+            if let Some((key, value)) = function_exporters.get(name) {
+                match value.as_slice() {
+                    [] => unreachable!(),
+                    [exporter] => {
+                        resolved.insert(key, *exporter);
+                    }
+                    _ => {
+                        duplicates.push((metadata.name, *key, value.as_slice()));
+                    }
+                }
+            } else {
+                missing.push((
+                    metadata.name,
+                    Export {
+                        key: ExportKey {
+                            name,
+                            ty: Type::Function(FunctionType {
+                                parameters: Vec::new(),
+                                results: Vec::new(),
+                            }),
+                        },
+                        flags: 0,
+                    },
+                ));
+            }
+        }
+    }
+
+    (resolved, missing, duplicates)
+}
+
+/// Recursively add a library (represented by its offset) and its dependency to the specified set, maintaining
+/// topological order (modulo cycles).
+fn topo_add<'a>(
+    sorted: &mut IndexSet<usize>,
+    dependencies: &HashMap<usize, HashSet<usize>>,
+    element: usize,
+) {
+    let empty = &HashSet::new();
+    let deps = dependencies.get(&element).unwrap_or(empty);
+
+    // First, add any dependencies which do not depend on `element`
+    for &dep in deps {
+        if !(sorted.contains(&dep) || dependencies.get(&dep).unwrap_or(empty).contains(&element)) {
+            topo_add(sorted, dependencies, dep);
+        }
+    }
+
+    // Next, add the element
+    sorted.insert(element);
+
+    // Finally, add any dependencies which depend on `element`
+    for &dep in deps {
+        if !sorted.contains(&dep) && dependencies.get(&dep).unwrap_or(empty).contains(&element) {
+            topo_add(sorted, dependencies, dep);
+        }
+    }
+}
+
+/// Topologically sort a set of libraries (represented by their offsets) according to their dependencies, modulo
+/// cycles.
+fn topo_sort(count: usize, dependencies: &HashMap<usize, HashSet<usize>>) -> Result<Vec<usize>> {
+    let mut sorted = IndexSet::new();
+    for index in 0..count {
+        topo_add(&mut sorted, &dependencies, index);
+    }
+
+    Ok(sorted.into_iter().collect())
+}
+
+/// Analyze the specified library metadata, producing a map of transitive dependencies, where each library is
+/// represented by its offset in the original metadata slice.
+fn find_dependencies(
+    metadata: &[Metadata],
+    exporters: &HashMap<&ExportKey, (&str, &Export)>,
+) -> Result<HashMap<usize, HashSet<usize>>> {
+    let mut dependencies = HashMap::<_, HashSet<_>>::new();
+    let mut indexes = HashMap::new();
+    for (index, metadata) in metadata.iter().enumerate() {
+        indexes.insert(metadata.name, index);
```

Comment:
Could this, and insertions below, assert that they all succeed? (e.g. don't replace anything)